### PR TITLE
Qmix on gpu and with non-stacked-obs environment state support

### LINKED ISSUE #5732: [rllib] Qmix does not use GPU, also QMixTorchPolicy inherits from Policy instead of TorchPolicy
### System information
- **OS Platform and Distribution (e.g., Linux Ubuntu 16.04)**: Linux Ubuntu 19.04
- **Ray installed from (source or binary)**: source
- **Ray version**: 8903bcd
- **Python version**: 3.7.4
- **Exact command to reproduce**: N/A

### Describe the problem
Qmix (implemented in Pytorch) does not use the GPU. 

It looks like TorchPolicy has some code to put the NN model(s) on the GPU and maybe do gradient stuff on the GPU as well, but QMixTorchPolicy inherits from the parent class Policy instead of TorchPolicy so it doesn't get that code (also TorchPolicy's `__init__` wants a `lock` but QMixTorchPolicy doesn't seem to use locks at all so it's not a straightforward port).

### Source code / logs
https://github.com/ray-project/ray/blob/master/rllib/agents/qmix/qmix_policy.py

### PR DESCRIPTION
**Pull request by mawright**
## Why are these changes needed?
Makes Qmix run on the GPU if it is available. I copied the basic method from TorchPolicy. 

Also adds support for an environment state that isn't the agents' stacked observations.

## Related issue number

Closes #5732 

## Checks

- [ x] I've run `scripts/format.sh` to lint the changes in this PR.
- [ x] I've included any doc changes needed for https://ray.readthedocs.io/en/latest/.

### CODE REVIEW

File: rllib/agents/qmix/qmix_policy.py [hunk 1]
```diff
@@ -45,8 +48,17 @@ def __init__(self,
...
+    def forward(self,
+                rewards,
+                actions,
+                terminated,
+                mask,
+                obs,
+                next_obs,
+                action_mask,
+                next_action_mask,
+                state=None,
```
Comment (ericl) [inline rllib/agents/qmix/qmix_policy.py]: To avoid confusing this with RNN state, could we call this `global_obs` here and elsewhere?

File: rllib/agents/qmix/qmix_policy.py [hunk 2]
```diff
@@ -146,6 +168,11 @@ def __init__(self, obs_space, action_space, config):
         self.n_agents = len(obs_space.original_space.spaces)
         self.n_actions = action_space.spaces[0].n
         self.h_size = config["model"]["lstm_cell_size"]
+        self.has_state = False
```
Comment (ericl) [inline rllib/agents/qmix/qmix_policy.py]: `has_global_obs`?

File: rllib/agents/qmix/qmix_policy.py [hunk 3]
```diff
@@ -160,6 +187,11 @@ def __init__(self, obs_space, action_space, config):
                     (self.n_actions, ), mask_shape))
             self.has_action_mask = True
             self.obs_size = _get_size(agent_obs_space.spaces["obs"])
+            if "state" in space_keys:
```
Comment (ericl) [inline rllib/agents/qmix/qmix_policy.py]: Can we make this a constant at the top of this file and add a comment?

File: rllib/agents/qmix/qmix_policy.py [hunk 4]
```diff
@@ -226,133 +257,170 @@ def compute_actions(self,
...
-            masked_q_values[avail == 0.0] = -float("inf")
-            # epsilon-greedy action selector
-            random_numbers = th.rand_like(q_values[:, :, 0])
-            pick_random = (random_numbers < self.cur_epsilon).long()
-            random_actions = Categorical(avail).sample().long()
-            actions = (pick_random * random_actions +
-                       (1 - pick_random) * masked_q_values.max(dim=2)[1])
-            actions = actions.numpy()
-            hiddens = [s.numpy() for s in hiddens]
+        with self.lock:
```
Comment (ericl) [inline rllib/agents/qmix/qmix_policy.py]: I guess the lock is needed for APEX mode only, is that right?

File: rllib/agents/qmix/qmix_policy.py [hunk 5]
```diff
@@ -125,7 +147,7 @@ def forward(self, rewards, actions, terminated, mask, obs, next_obs,
         return loss, mask, masked_td_error, chosen_action_qvals, targets
 
 
-class QMixTorchPolicy(Policy):
+class QMixTorchPolicy(TorchPolicy):
```
Comment (ericl) [inline rllib/agents/qmix/qmix_policy.py]: We should probably call `TorchPolicy.__init__` if extending TorchPolicy. I'm not sure if extending it is necessary though, since QMix doesn't really use the common facilities provided by TorchPolicy.

**Review by ericl — COMMENTED** [review]
Thanks, I left a few comments. Could you also modify twostep_game to include the new global state/obs field? That example is automatically run in run_rllib_tests.sh with the --run=QMIX arg, so it will test that it runs successfully.

File: rllib/agents/qmix/qmix_policy.py (hunk 4 above)
Comment (mawright) [reply rllib/agents/qmix/qmix_policy.py]: I was just copying what was done in the TorchPolicy class. Are the locks needed in the read operations?

File: rllib/agents/qmix/qmix_policy.py (hunk 1 above)
Comment (mawright) [reply rllib/agents/qmix/qmix_policy.py]: how about env_state?

**Comment by ericl** [comment]
I don't think you need locks unless it's crashing. Torch has also gotten a
lot better about thread safety.

**Comment by mawright** [comment]
Should get_weights() and set_weights() also operate on the "mixer" critic neural network?

**Comment by mawright** [comment]
That is, are get_weights() and set_weights() ever used when we want to transfer the critic network between processes and not just the actor network?

**Comment by ericl** [comment]
I think it's a good idea since it may be used for checkpointing. The
performance overhead should be pretty minimal, so better to keep it safe
and transfer everything.

Eric

**Comment by mawright** [comment]
@ericl I'm pretty sure the test failures are not my fault. I see them arising in code I didn't touch (and mostly in Tensorflow code that probably broke with Tensorflow 2.0)

**Review by ericl — APPROVED** [review]

File: rllib/agents/qmix/model.py [hunk 6]
```diff
@@ -31,7 +31,7 @@ def get_initial_state(self):
 
     @override(TorchModelV2)
     def forward(self, input_dict, hidden_state, seq_lens):
-        x = F.relu(self.fc1(input_dict["obs_flat"].float()))
+        x = F.relu(self.fc1(input_dict["obs"].float()))
```
Comment (ericl) [inline rllib/agents/qmix/model.py]: Would it also work to be obs_flat? That would let it handle complex observation spaces which wouldn't be a flat vector.

Comment (mawright) [reply rllib/agents/qmix/model.py]: Unless I am mistaken, obs_flat includes all of the elements in the agent's obs dict, flattened and concatted, which, if we are putting the environment global state in the agents' obs dict, means that the agent's policy gets that info. We could make it so there is an added env state obs dict alongside the grouped agent obs dicts that come out of GroupAgentsWrapper (i.e., the env's observation space is a (potentially nested) dict with keys eg "env_state", "group_1", ...). That would require some refactoring of that wrapper and probably some other parts of the code.

Comment (mawright) [reply rllib/agents/qmix/model.py]: Actually, I just realized that if you look here: https://github.com/ray-project/ray/blob/785670bc18a8595219c96e9512192922fafcf510/rllib/agents/qmix/qmix_policy.py#L426-L443

, which is the only part of qmix_policy.py where the NN's forward pass is called, there is an assumption here, before forward() is entered, that the observation tensor for each agent is a vector.

Comment (ericl) [reply rllib/agents/qmix/model.py]: I think obs flat and obs are equivalent then, since you pass in only the agent obs to model()? It should not be able to find the original obs that has the env state.

Comment (mawright) [reply rllib/agents/qmix/model.py]: I think that is correct here. It looks like it was done this way in the original writing of qmix_policy where a new dict that is just the "obs" tensor is passed into the NN model so that the "action_mask" tensor would not be passed in in obs_flat.
